torch/_functorch/_aot_autograd/runtime_wrappers.py

```diff
@@ -1891,15 +1891,63 @@ def _backward_prologue(ctx, *flat_args):
                             "The grad inputs should be same number as forward output tangents"
                         )
 
-                    flat_processed_tangents = list(
+                    _num_mutated_inp_runtime = (
+                        CompiledFunction.metadata.num_mutated_inp_runtime_indices
+                    )
+
+                    flat_processed_tangents: List[torch.Tensor] = []
+                    if _num_mutated_inp_runtime > 0:
+
+                        def subclass_zero_leaves(
+                            meta: SubclassCreationMeta,
+                            output_append_list: List[torch.Tensor],
+                        ):
+                            for attr_meta in meta.attrs.values():
+                                if isinstance(attr_meta, SubclassCreationMeta):
+                                    subclass_zero_leaves(attr_meta, output_append_list)
+                                else:
+                                    assert attr_meta.shape_dtype_device is not None
+                                    output_append_list.append(
+                                        torch.zeros(
+                                            attr_meta.shape_dtype_device[0],
+                                            dtype=attr_meta.shape_dtype_device[1],
+                                            device=attr_meta.shape_dtype_device[2],
+                                        )
+                                    )
+
+                        for t, m in zip(
+                            tangents[:_num_mutated_inp_runtime],
+                            CompiledFunction.metadata.subclass_tangent_meta[
+                                :_num_mutated_inp_runtime
+                            ],
+                        ):
+                            if isinstance(m, SubclassCreationMeta):
+                                if type(t) == torch.Tensor:
```

Comment:
I think this is technically wrong: this assumes that if we guessed the tangent was a subclass, and it was actually a plain tensor (and it maps to a mutated fw input), then it **must** be a tensor of zeros. But that does not have to be true: it's possible that someone later used the (updated) input in some autograd-related compute, but the subclass disappeared when running the backward.

Rather than making a guess about adding extra zero tensor arguments, I think we should either:

(1) "guess better": today, we make the guess when generating our FakeTensor tangents for tracing that if any of the fw_outs are subclasses, their tangents will also be subclasses. At the time that we generate this guess, we could tweak it so that if a `fw_out` corresopnds to an `updated_input` that we are returning, then we can guess that it will be a plain tensor

(2) if this is too much of a pain, we can also just give up on this case for now

Reply:
Thanks! I agree, guessing better is much cleaner. updated the PR with this logic.